### PR TITLE
ruby: Allow parallel startup of PhantomJS/IE/Safari drivers

### PR DESCRIPTION
**Pull request by p0deje**
This extends 437725c to PhantomJS because any external process starting
with specific port suffers the same race condition.

Basically, we apply the behavior we have for Firefox (lock socket, start firefox, unlock socket) to any external process that we pass port to. 437725c did that for Chrome, this pull request - for PhantomJS.

I talked to @jarib about this and he advised first to propose this to the team. That kind of changes should be applied to all bindings, so it makes sense to discuss it first and I didn't manage to get attention in IRC. @barancev @titusfortner @lukeis @AutomatedTester @andreastt What do you think?

If it makes sense to do that, I'll need to apply this behavior to IE driver in Ruby (if it makes sense - I have no idea about IEDriverServer and how it works in parallel).

### CODE REVIEW

File: rb/lib/selenium/webdriver/phantomjs/service.rb [hunk 1]
```diff
@@ -100,10 +102,22 @@ def create_process(args)
             process.io.stdout = process.io.stderr = File.new(Platform.null_device, 'w')
           end
 
-          process
+          @process.start
+        end
+
+        def connect_until_stable
+          @socket_poller = SocketPoller.new @host, @port, START_TIMEOUT
```
Comment (titusfortner) [inline rb/lib/selenium/webdriver/phantomjs/service.rb]: Where else is this used that we need to make this an instance variable?

**Comment by titusfortner** [comment]
Matches up with chrome; I haven't used Phantomjs to test it. @jimevans is probably the person to ask about the IEDriver.

File: rb/lib/selenium/webdriver/phantomjs/service.rb (hunk 1 above)
Comment (p0deje) [inline rb/lib/selenium/webdriver/phantomjs/service.rb]: True, it should be local variable.

File: rb/lib/selenium/webdriver/phantomjs/service.rb [hunk 2]
```diff
@@ -100,10 +102,22 @@ def create_process(args)
...
+
+          unless @socket_poller.connected?
+            raise Error::WebDriverError, "unable to connect to phantomjs @ #{uri} after #{START_TIMEOUT} seconds"
```
Comment (jarib) [inline rb/lib/selenium/webdriver/phantomjs/service.rb]: This will fail with 'undefined local variable or method "uri"'.

**Comment by bootstraponline** [comment]
:+1:

File: rb/lib/selenium/webdriver/phantomjs/service.rb (hunk 1 above)
Comment (p0deje) [inline rb/lib/selenium/webdriver/phantomjs/service.rb]: Fixed in all other services as well

**Comment by p0deje** [comment]
@titusfortner Please, take a look. I've added basic specs which ensure that multiple instances of driver can be started concurrently - each of these specs fails if `SocketLock` is removed from service implementation. I will extend the spec to cover Edge once I have Windows development environment set up. Not sure what to do with IE and Safari (looks like will have to research).

**Comment by titusfortner** [comment]
Yes, with specs, good.

As for your windows environment, I've set up and re-set up Windows on a VM over a dozen times the past week. To make it easier, I played around with Jim's Windows Selenium Setup. This gets everything almost set up for running Ruby tests: https://github.com/titusfortner/windows-selenium-dev-setup
A bunch is hard coded to current versions, and a bunch of stuff has to be done manually at the end because I couldn't figure out how to keep it from exiting or forcing it to restart. It does at least show what all needs to be added to get things working for us.

**Comment by p0deje** [comment]
@titusfortner @lukeis 

Please take a look, this now allows to start PhantomJS, IE and Safari in parallel without any issues. I've created separated issues describing particular problems of parallel startup with scripts to reproduce (they basically duplicate these specs). I'd like to add a spec for Edge as well, but I don't have a Win 10 VM to properly test it. I've also tested everything thoroughly and verified specs are both failing before fixing and passing after:

1. PhantomJS: 1.9.8 on OS X, 2.0 on Windows 7.
2. IE: 9.0 on Windows 7 (I don't see how other versions may affect it).
3. Safari: 9 on OS X El Capitain.

If you're fine with these changes, I'm going to merge and close related issues and start working on porting this to Python bindings (that's going to be fun).

**Comment by titusfortner** [comment]
This looks good! I've been doing a bunch of spec testing recently on Mac, Linux & Win10. I'll run these through those paces in the next couple days, and make sure it works for Marionette as well. I finally have the latest build of Win10 installed, so I can test an implementation for Edge as well if you want to add any more code for that.

**Comment by p0deje** [comment]
If you could try to add specs for Marionette and Edge just like I've added for IE/PhantomJS and see if it works, that'd be really great!

**Comment by titusfortner** [comment]
@p0deje - so, [this commit](https://github.com/titusfortner/selenium/commit/71aed87a33b014b923a719153f500c59671e3893) opens up 5 instances of wires, and 5 instances of Firefox with Marionette, but times out without making the driver call. It works if I change the 5.times to 1.times, but I don't know what is going on that keeps it from working otherwise, or whose fault it is....

I haven't had a chance to test it on edge/ie, that's next.

**Comment by p0deje** [comment]
@titusfortner I'll try with Wires to see how it works.

**Comment by p0deje** [comment]
@titusfortner I'm getting error "Your Firefox profile cannot be loaded. It may be missing or inaccessible" when drivers are started in parallel and I believe this is some bug in Marionette/Wires. I'll investigate and report it.

Do you agree we can merge this pull request even without spec for Wires?

**Comment by titusfortner** [comment]
Yeah it doesn't work for either Marionette on any platform or Edge on Windows 10. It works on all other combinations, so I think this is good to merge in for everything else. Thanks!

**Pull request merged**